### DISCUSSION TITLE
session.merge() with condition

### DISCUSSION THREAD

**Discussion by yishang11** [sqlalchemy/sqlalchemy#10631] [Usage Questions]
Hi I am still using sqlalchemy 1.4 and sqlite3, and got stuck when combining a bunch of databases sharing the same schema into one database. For example I have `source_db` having table A with 1 row, `target_db` having table A with 3 row, and I want to insert the rows from source to target. If I do `session.merge()` it would replace row 1 in `target_db` instead. 
Below are two paths I am trying but couldn't figure out:
- Is it possible to do merging with conditions so it would always insert instead of trying to resolve conflicts?
- Is it possible to grep all the properties of the source row and insert a new row using all these properties, in an automatic fashion? 
Thank you!

**Comment by zzzeek** [comment]
in order to have an object from a session, and to INSERT it somewhere else like it's brand new row, use [make_transient](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.make_transient), then add it to the other session using `session.add()`.  It will emit an INSERT unconditionally.   you will want to "null out" the primary key field if you are using an auto-incrementing integer primary key column.

**Reply by yishang11** [reply]
Hi zzzeek, thank you for your response! Now I have a follow up question.
In order to use `session.add()`, I needed to expire any loaded relationship of the src obj, otherwise I get 
```
sqlalchemy.exc.InvalidRequestError: Object '<Data at 0x7f0578b5ffa0>' is already attached to session '1' (this is '2')
```
I wanted to add an assertion that all lazy-loaded relationship has been expired, but couldn't figure out how to do this.
I tried `inspect(src_obj).unloaded` which would tell me unloaded relationships, but I need to do something like `assert inspect(src_obj).loaded == {}`. The inspect obj has only `unloaded` attribute, not `loaded`.
Thank you!

**Reply by zzzeek** [reply]
you can expire all relationships like this:

```py
session.expire(obj, inspect(obj).mapper.relationships.keys())
```
